File: src/main/java/com/kingdom/util/computercardaction/PromoComputerCardActionHandler.java
package com.kingdom.util.computercardaction;

import com.kingdom.model.Card;
import com.kingdom.model.CardAction;
import com.kingdom.model.Game;
import com.kingdom.model.Player;
import com.kingdom.model.computer.ComputerPlayer;
import com.kingdom.util.cardaction.CardActionHandler;

import java.util.ArrayList;
import java.util.List;

public class PromoComputerCardActionHandler {
    public static void handleCardAction(CardAction cardAction, ComputerPlayer computer) {

        Game game = computer.getGame();
        Player player = computer.getPlayer();

        String cardName = cardAction.getCardName();
        int type = cardAction.getType();
        if (cardName.equals("Black Market")) {
            if (type == CardAction.TYPE_YES_NO) {
                CardActionHandler.handleSubmittedCardAction(game, player, null, "no", null, -1);
            } else if (type == CardAction.TYPE_CHOOSE_CARDS) {
                List<Integer> cardIds = new ArrayList<Integer>();
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            } else if (type == CardAction.TYPE_CHOOSE_IN_ORDER) {
                List<Integer> cardIds = new ArrayList<Integer>();
                for (Card card : cardAction.getCards()) {
                    cardIds.add(card.getCardId());
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            }
        } else if (cardName.equals("Envoy")) {
            //todo determine which card would be best to get
            Card cardToDiscard = computer.getHighestCostCard(cardAction.getCards(), false);
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToDiscard.getCardId());
            CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Governor")) {
            if (type == CardAction.TYPE_CHOICES) {
                String choice;
                if (computer.getNumCardsWorthTrashing(player.getHand()) > 0) {
                    choice = "trash";
                } else if (computer.getGoldsBought() == 0) {
                    choice = "money";
                } else {
                    choice = "cards";
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null, null, choice, -1);
            } else if (type == CardAction.TYPE_TRASH_UP_TO_FROM_HAND) {
                List<Integer> cardIds;
                if (computer.getNumCardsWorthTrashing(cardAction.getCards()) > 0) {
                    cardIds = computer.getCardsToTrash(cardAction.getCards(), 1);
                } else {
                    cardIds = new ArrayList<Integer>();
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            } else if (type == CardAction.TYPE_GAIN_CARDS_FROM_SUPPLY) {
                Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
                List<Integer> cardIds = new ArrayList<Integer>();
                cardIds.add(cardToGain.getCardId());
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            }
        } else if (cardName.equals("Walled Village")) {
            if (type == CardAction.TYPE_YES_NO) {
                CardActionHandler.handleSubmittedCardAction(game, player, null, "yes", null, -1);
            } else {
                List<Integer> cardIds = new ArrayList<Integer>();
                for (Card card : cardAction.getCards()) {
                    cardIds.add(card.getCardId());
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            }
        } else {
            throw new RuntimeException("Promo Card Action not handled for card: " + cardAction.getCardName() + " and type: " + type);
        }
    }
}

File: src/main/kotlin/com/kingdom/util/computercardaction/PromoComputerCardActionHandler.kt
package com.kingdom.util.computercardaction

import com.kingdom.model.OldCardAction
import com.kingdom.model.computer.ComputerPlayer
import com.kingdom.util.cardaction.CardActionHandler
import java.util.*

object PromoComputerCardActionHandler {
    fun handleCardAction(oldCardAction: OldCardAction, computer: ComputerPlayer) {

        val game = computer.game
        val player = computer.player

        val cardName = oldCardAction.cardName
        val type = oldCardAction.type

        when (cardName) {
            "Black Market" -> when (type) {
                OldCardAction.TYPE_YES_NO -> CardActionHandler.handleSubmittedCardAction(game, player, null!!, "no", null, -1)
                OldCardAction.TYPE_CHOOSE_CARDS -> {
                    val cardNames = ArrayList<String>()
                    CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
                }
                OldCardAction.TYPE_CHOOSE_IN_ORDER -> {
                    val cardNames = ArrayList<String>()
                    for (card in oldCardAction.cards) {
                        cardNames.add(card.name)
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
                }
            }
            "Envoy" -> {
                //todo determine which card would be best to get
                val cardToDiscard = computer.getHighestCostCard(oldCardAction.cards, false)
                val cardNames = ArrayList<String>()
                cardNames.add(cardToDiscard!!.name)
                CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
            }
            "Governor" -> when (type) {
                OldCardAction.TYPE_CHOICES -> {
                    val choice = if (computer.getNumCardsWorthTrashing(player.hand) > 0) {
                        "trash"
                    } else if (computer.goldsBought == 0) {
                        "money"
                    } else {
                        "cards"
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, choice, -1)
                }
                OldCardAction.TYPE_TRASH_UP_TO_FROM_HAND -> {
                    val cardNames = if (computer.getNumCardsWorthTrashing(oldCardAction.cards) > 0) {
                        computer.getCardsToTrash(oldCardAction.cards, 1)
                    } else {
                        ArrayList()
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
                }
                OldCardAction.TYPE_GAIN_CARDS_FROM_SUPPLY -> {
                    val cardToGain = computer.getHighestCostCard(oldCardAction.cards)
                    val cardNames = ArrayList<String>()
                    cardNames.add(cardToGain!!.name)
                    CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
                }
            }
            "Walled Village" -> when (type) {
                OldCardAction.TYPE_YES_NO -> CardActionHandler.handleSubmittedCardAction(game, player, null!!, "yes", null, -1)
                else -> {
                    val cardNames = oldCardAction.cards.map { it.name }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardNames, null, null, -1)
                }
            }
            else -> throw RuntimeException("Promo Card Action not handled for card: " + oldCardAction.cardName + " and type: " + type)
        }
    }
}
